File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.cpp/core/src/impl/RemoveEntityCommandImpl.cpp
#include "RemoveEntityCommandImpl.h"

#include "Entity.h"

namespace commands {

RemoveEntityCommandImpl::RemoveEntityCommandImpl() {
}

void RemoveEntityCommandImpl::execute() {
    auto entity = getEntity();
    EntityReference reference = getEntityToRemove().get();
    entity->removeFromCollection(getPropertyName(), reference);
}

void RemoveEntityCommandImpl::undo() {

    auto entity = getEntity();
    EntityReference reference = getEntityToRemove().get();
    entity->addToCollection(getPropertyName(), reference);
}

void RemoveEntityCommandImpl::redo() {
	execute();
}

}

File: simulators/de.unistuttgart.iste.sqa.mpw.framework.java/core/src/main/java/de/unistuttgart/iste/sqa/mpw/framework/commands/impl/RemoveEntityCommandImpl.java
package de.unistuttgart.iste.sqa.mpw.framework.commands.impl;

import de.unistuttgart.iste.sqa.mpw.framework.commands.impl.utils.CommandUtils;
import de.unistuttgart.iste.sqa.mpw.framework.commands.RemoveEntityCommand;

/**
 * Note: position may be different after undo.
 */
public class RemoveEntityCommandImpl extends RemoveEntityCommand {

	@Override
	public void execute() {
		try {
			final var entity = getEntity();
			final var method = CommandUtils.findRemoveMethod(entity, getPropertyName());
			method.invoke(entity, getEntityToRemove());
		} catch (final Exception e) {
			throw new RuntimeException("Error: could not execute RemoveEntityCommand '" + getPropertyName() + "' on entity " + getEntity(), e);
		}
	}

	@Override
	public void undo() {
		try {
			final var entity = getEntity();
			final var method = CommandUtils.findAddMethod(entity, getPropertyName());
			method.invoke(entity, getEntityToRemove());
		} catch (final Exception e) {
			throw new RuntimeException("Error: could not execute RemoveEntityCommand '" + getPropertyName() + "' on entity " + getEntity(), e);
		}
	}

	@Override
	public void redo() {
		execute();
	}

}
